src/pthread_esm_startup.mjs

```diff
@@ -0,0 +1,48 @@
+/**
+ * @license
+ * Copyright 2025 The Emscripten Authors
+ * SPDX-License-Identifier: MIT
+ */
+
+// This file is used as the initial script loaded into pthread workers when
+// running in WASM_ESM_INTEGRATION mode.
+// The point of this file is to delay the loading the main program module
+// until the wasm memory has been received via postMessage.
+
+#if RUNTIME_DEBUG
+console.log("Running pthread_esm_startup");
+#endif
+
+#if ENVIRONMENT_MAY_BE_NODE
+// Create as web-worker-like an environment as we can.
+var worker_threads = await import('worker_threads');
+global.Worker = worker_threads.Worker;
+var parentPort = worker_threads['parentPort'];
+parentPort.on('message', (msg) => global.onmessage?.({ data: msg }));
+Object.assign(globalThis, {
+  self: global,
+  postMessage: (msg) => parentPort['postMessage'](msg),
+});
+#endif
+
+self.onmessage = async (msg) => {
+#if RUNTIME_DEBUG
+  console.log('pthread_esm_startup', msg.data.cmd);
+#endif
+  if (msg.data.cmd == 'load') {
+    // Until we initialize the runtime, queue up any further incoming messages.
+    let messageQueue = [msg];
+    self.onmessage = (e) => messageQueue.push(e);
```

Comment:
I'm confused.

1. Isn't this queuing also needed *before* "load" arrives? It seems like it applies after load, for things in between load and some future event..?
2. It looks like we process this queue right after the await - to be sure I follow, is the queue only for events arriving during the await?
3. Do we not need to undo the trampling of `self.onmessage` at the end?

Reply:
I added some more comments that should explain what is doing on there more precisely.